### DISCUSSION TITLE
Should Record and Verify tasks be decoupled from Test task?

### DISCUSSION THREAD

**Discussion by sampengilly** [cashapp/paparazzi#1127] [Ideas]
In evaluating Paparazzi, one thing that hasn't been super clear to me is how it operates in a project where tests making use of the paparazzi test rule are just one of many different unit tests. Particularly with regards to build performance and skipping work.

From a read of the Gradle plugin, it appears that the `verify` and `record` tasks extend from the underlying Test task, meaning they run the test suite themselves. Now Gradle will probably be smart enough to skip repeated work if you run `:test` followed by one of the paparazzi tasks. But it's not entirely clear that this would be the case, especially since it looks like the behaviour of the test rule changes depending on which gradle task is being run.

Wouldn't it make more sense for the Paparazzi test rule to output snapshots/reports to the build directory as part of a normal gradle `:test` run, and for the record and verify tasks to depend on those outputs to do the image comparison/overwriting as an operation that follows/depends on the test run rather than as part of a potentially second test run?

**Comment by geoff-powell** [comment]
That is an interesting approach. One issue with that is we can't "fail fast" on the verify step if verification happens after test task runs.

I need to think on this some more.
